3557 langchain4j in jee mp app (#3674)
OpenLiberty/blogs: posts/2024-03-28-open-liberty-with-langchain4j-example.adoc

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
additional_authors:
- name: Grace Jansen
  github: https://github.com/GraceJansen
  image: https://avatars0.githubusercontent.com/GraceJansen
---
= Run LangChain4j in Jakarta EE and MicroProfile application on Open Liberty
Gilbert Kwan <https://github.com/gkwan-ibm>
:imagesdir: /
:url-prefix:
:url-about: /

:example-location: https://github.com/langchain4j/langchain4j-examples/tree/main/jakartaee-microprofile-example

Artificial Intelligence (AI) is an exciting and disruptive field that is already transforming businesses, and even entire industries, by enabling automation, improving decision-making and unlocking new insights from data. With the rise in large language models, or LLMs, such as ChatGPT, over the last few years there has been a significant shift in the performance of AI and its potential to drive enterprise value. So, how will this impact on software development and the creation of cloud-native Java applications for enterprises?

> In this blog we'll explore what LLMs are, frameworks that can be used to simplify their use in Java applications and how to get started using them in a Jakarta EE/MicroProfile app:

Reviewer: ```suggestion
In this post, we'll explore what LLMs are and how to use them in Java applications. We'll also get started using them in an example Jakarta EE/MicroProfile app.
```

* <<what_Are_LLMs, Find out what LLMs are>>
* <<Java_LLMs, Explore app frameworks that simplify the use of LLMs>>

> * <<using_Langchain4j, Discover how LangChain can be used in Java applications>>

Reviewer: ```suggestion
* <<using_Langchain4j, Discover how to use LangChain in Java applications>>
```

* <<tryout, Try out the jakartaee-microprofile-example application>>
* <<how_app_work, Learn how this sample application works>>

[#what_Are_LLMs]
== What are LLMs?

> A language model is a model of natural language based on probabilities. They are able to generate probabilities of a series of words together. https://www.ibm.com/topics/large-language-models[Large language models (LLMs)] are simply langauge models that are categorized by their large size.They are trained on immense amounts of data, possibly billions of paramters, using self-supervised and semi-supervised learning techniques. This makes them capable of understanding and generating natural language and other types of content to perform a wide range of tasks.

Reviewer: ```suggestion
A language model is a model of natural language based on probabilities. They are able to generate probabilities of a series of words together. https://www.ibm.com/topics/large-language-models[Large language models (LLMs)] are simply langauge models that are categorized by their large size.They are trained on immense amounts of data, possibly billions of parameters, using self-supervised and semi-supervised learning techniques. This training enables them to generate natural language and other types of content to perform a wide range of tasks.
```

Reviewer: I'd be wary of claiming LLMs "understand" things

> See the introductory video below for more information on LLMs, including what they are, how they work and business applications.

Reviewer: ```suggestion
See the introductory video below for more information on LLMs, including what they are, how they work, and their business applications.
```

video::5sLYAQS9sWQ[youtube, width="560", height="315", align="center"]

> LLMs can be found in the service offerings of almost all of the major cloud service providers, e.g. IBM offers models through it's https://www.ibm.com/watsonx[watsonx] services, https://azure.microsoft.com/en-us/solutions/ai[Microsoft Azure] offers LLMs like Llama 2 and OpenAI GPT-4, and  https://aws.amazon.com/bedrock/[Amazon Bedrock] offers models from a range of AI companies.

Reviewer: ```suggestion
You can find LLMs in the service offerings of almost all of the major cloud service providers. For example, IBM offers models through it's https://www.ibm.com/watsonx[watsonx] services, https://azure.microsoft.com/en-us/solutions/ai[Microsoft Azure] offers LLMs like Llama 2 and OpenAI GPT-4, and  https://aws.amazon.com/bedrock/[Amazon Bedrock] offers models from a range of AI companies.
```

[#Java_LLMs]
== How can we take advantage of LLMs in Java applications?

> Developing applications powered by large language models isn't always an easy task. Integrating AI/LLM capabilities into an application can be challenging. To make this easier, LangChain, an open source framework was developed in 2022 to help streamline the process of creating generative AI apps.

Reviewer: ```suggestion
Integrating AI/LLM capabilities into an application can be challenging. The open source LangChain framework was developed in 2022 to help streamline the process of creating generative AI apps.
```

LangChain provides tools and abstractions to improve the customization, accuracy, and relevancy of the information the models generate. For example, developers can use LangChain components to build new prompt chains or customize existing templates. LangChain also includes components that allow LLMs to access new data sets without retraining and organizes the large quantities of data these models require so that they can be accessed with ease.

> Although LangChain is primarily available in Python and JavaScript/TypeScript versions, there are options available to use LangChain in Java applications through community projects like https://github.com/langchain4j/langchain4j[LangChain4j]. LangChain4j APIs provide the ability to intergate LLMs into your Java application for different AI platforms, such as OpenAI and Hugging Face, etc.

Reviewer: ```suggestion
Although LangChain is primarily available in Python and JavaScript/TypeScript versions, options are available to use LangChain in Java applications through community projects like https://github.com/langchain4j/langchain4j[LangChain4j]. LangChain4j APIs provide the ability to integrate LLMs into your Java application for different AI platforms, such as OpenAI and Hugging Face, etc.
```

image::/img/blog/langchain4j.png[langchain4j GitHub README image,width=70%,align="center"]

[#using_Langchain4j]
== How to use LangChain4j in a Jakarta EE and MicroProfile application

Langchain4j has a very useful open source https://github.com/langchain4j/langchain4j-examples[langchain4j-examples] GitHub repositry where it stores example applications. However, although this repo provided useful examples of using LangChain4j in general Java apps, we could not find any examples showcasing how you could experience these AI technologies in a Jakarta EE/MicroProfile based application. So, we decided to build one ourselves called `jakartaee-microprofile-example` which can now be found in this https://github.com/langchain4j/langchain4j-examples/tree/main/jakartaee-microprofile-example[langchain4j-examples] GitHub repository. This demo application highlights how to use LangChain4j APIs in an application using Jakarta EE and MicroProfile on Open Liberty.

[#tryout]
== Try out the jakartaee-microprofile-example application

> To see how you could apply LangChain4j to your own Jakarta EE and/or MicroProfile application, check out this example project for yourself.

Reviewer: ```suggestion
To see how you could apply LangChain4j to your own Jakarta EE and MicroProfile applications, check out this example project for yourself.
```

[#pre-reqs]

> === Pre-requisites

Reviewer: ```suggestion
=== Prerequisites
```

> Before you clone the application to your machine, install JDK 17 and ensure that your `JAVA_HOME` environment variable is set.  You can make use of the https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtime] as your chosen Java runtime. This runtime offers performance benefits from deep technology investment in projects such as Eclipse OpenJ9 and is available across a wide variety of hardware and software platforms. To find out more about IBM Semeru Runtime, see https://openliberty.io/blog/2022/08/19/open-liberty-semeru-performance.html[Open Liberty and Semeru Runtimes, cloud-native performance that matters].

Reviewer: ```suggestion
Before you clone the application to your machine, install JDK 17 and ensure that your `JAVA_HOME` environment variable is set.  You can use the https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtime] as your chosen Java runtime. This runtime offers performance benefits from deep technology investment in projects such as Eclipse OpenJ9 and is available across a wide variety of hardware and software platforms. To find out more about IBM Semeru Runtime, see https://openliberty.io/blog/2022/08/19/open-liberty-semeru-performance.html[Open Liberty and Semeru Runtimes, cloud-native performance that matters].
```

The application uses Hugging Face. You need to get a Hugging Face API Key:

* Sign up and login to https://huggingface.co

> * Go to Access Tokens by https://huggingface.co/settings/tokens

Reviewer: ```suggestion
* Go to the link:https://huggingface.co/settings/tokens[Access Tokens page] 
```

> * Create a new access token with "read" role

Reviewer: ```suggestion
* Create a new access token with `read` role
```

> To access the repository remotely you'll need to install https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[Git] if you haven't already. You can clone the `langchain4j-examples` GitHub repository to your local machine by running:

Reviewer: ```suggestion
To access the repository remotely, install https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[Git] if you haven't already. You can clone the `langchain4j-examples` GitHub repository to your local machine by running the following command:
```

[source]
----
git clone https://github.com/langchain4j/langchain4j-examples.git
----

[#env_setup]

> === Environment Set Up

Reviewer: ```suggestion
=== Environment setup
```

To run the application, navigate to the `jakartaee-microprofile-example` directory:

[source]
----
cd langchain4j-examples/jakartaee-microprofile-example
----

> and set the environment variables:

Reviewer: ```suggestion
Set the following environment variables:
```

[source]
----
export JAVA_HOME=<your Java 17 home path>
export HUGGING_FACE_API_KEY=<your Hugging Face read token>
----

[#start_app]
=== Start the application

> To start the application, use the provided Maven wrapper to run the https://openliberty.io/docs/latest/development-mode.html[Liberty dev mode]:

Reviewer: ```suggestion
To start the application, use the provided Maven wrapper to run Open Liberty in https://openliberty.io/docs/latest/development-mode.html[dev mode]:
```

[source]
----
./mvnw liberty:dev
----

After you see the following message, the application is ready:

----
************************************************************************
*    Liberty is running in dev mode.
*        Automatic generation of features: [ Off ]
*        h - see the help menu for available actions, type 'h' and press Enter.
*        q - stop the server and quit dev mode, press Ctrl-C or type 'q' and press Enter.
*
*    Liberty server port information:
*        Liberty server HTTP port: [ 9080 ]
*        Liberty server HTTPS port: [ 9443 ]
*        Liberty debug port: [ 7777 ]
************************************************************************
----

> To ensure the application has started successfully, you can run the tests by pressing the `enter/return` key from the command-line session. If the tests pass, you should see a similar output to the following example:

Reviewer: ```suggestion
To ensure the application started successfully, you can run the tests by pressing the `enter/return` key from the command-line session. If the tests pass, you should see output similar to the following example:
```

----
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running it.dev.langchan4j.example.ChatServiceIT
[INFO] ...
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.439 s...
[INFO] ...
[INFO] Running it.dev.langchan4j.example.ModelResourceIT
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.733 s...
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
----

[#access_app]
=== Access the application

Once the application is running, you can access it through a browser of your choice at http://localhost:9080/ and start experimenting with it.

image::/img/blog/langchain4j-example-chat-room.png[Chat Room of LangChain4j Jakarta EE and MicroProfile example,width=70%,align="center"]

You can type in any text that you want to chat with the AI agent. Here are some suggested messages:

* `What is MicroProfile?`
* `Which top 10 companies contribute MicroProfile?`

> * `any documentation?`

Reviewer: ```suggestion
* `Any documentation?`
```

Reviewer: this seems like too vague of a prompt to be effective but if it has been tested in context we can go with it

[source, java, role="no_copy"]
----
    public Assistant getAssistant() {
        ...
            HuggingFaceChatModel model = HuggingFaceChatModel.builder()
                .accessToken(HUGGING_FACE_API_KEY)
                .modelId(CHAT_MODEL_ID)
                .timeout(ofSeconds(TIMEOUT))
                .temperature(TEMPERATURE)
                .maxNewTokens(MAX_NEW_TOKEN)
                .waitForModel(true)
                .build();
            assistant = AiServices.builder(Assistant.class)
                .chatLanguageModel(model)
                .chatMemoryProvider(
                    sessionId -> MessageWindowChatMemory.withMaxMessages(MAX_MESSAGES))
                .build();
       ...
    }
----

Through the customized {example-location}/src/main/java/dev/langchain4j/example/chat/ChatAgent.java[`Assistant`] interface, the application can send messages to the LLM by its `chat()` method.

----
    interface Assistant {
       String chat(@MemoryId String sessionId, @UserMessage String userMessage);
    }
----

[#external_config]
=== Externalizing the configuration

> As the above code, to access the model, an API key is required. For security practice, it is not hard-coded in the code. The application externalizes the API key and the LangChain4j model properties with the MicroProfile Config feature that helps the application to run in different environments without code changes. You can learn more from the https://openliberty.io/docs/latest/external-configuration.html[External configuration of microservices] document.

Reviewer: ```suggestion
An API key is required to access the model. For security purposes, the key is not hard-coded in the code. The application externalizes the API key and the LangChain4j model properties with the MicroProfile Config feature that helps the application to run in different environments without code changes. You can learn more from the https://openliberty.io/docs/latest/external-configuration.html[External configuration of microservices] document.
```

----
hugging.face.api.key=set it by env variable
chat.model.id=NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO
chat.model.timeout=120
chat.model.max.token=200
chat.model.temperature=1.0
chat.memory.max.messages=20
----

[#communicate]
=== Communicating between the client and LLM

The application provides the interactive UI client for users to communicate with the LLM. Jakarta WebSocket enables two-way communication between the client and the `ChatService` service. Each client makes an HTTP connection to the service and send out the messages by the `send()` method.

See the {example-location}/src/main/webapp/chatroom.js[`src/main/webapp/chatroom.js`] file.
[source, java, role="no_copy"]
----
    const webSocket = new WebSocket('ws://localhost:9080/chat');
    ...
    function sendMessage() {
        ...
        var myMessage = document.getElementById('myMessage').value;
        ...
        webSocket.send(myMessage);
        ...
    }
----

> The service recieves the user messeages through the WebSocket `onMessage()` method, forward them to the LLM by calling the `ChatAgent.chat()` method, and then boardcast the LLM responsed answers back to the client session through the `sendObect()` method.

Reviewer: ```suggestion
The service recieves the user messeages through the WebSocket `onMessage()` method, forwards them to the LLM by calling the `ChatAgent.chat()` method, and then broadcasts the LLM answers back to the client session through the `sendObect()` method.
```

[source, java, role="no_copy"]
----
    @OnMessage
    public void onMessage(String message, Session session) {
        ...
        try {
            ...
            answer = agent.chat(sessionId, message);
        } catch (Exception e) {
            ...
        }

        try {
            session.getBasicRemote().sendObject(answer);
        } catch (Exception e) {
            e.printStackTrace();
        }

    }
----

[#enable_metrics]
=== Enabling metrics

To determine the performance and health of the application, it uses the MicroProfile Metrics feature to collect how much processing time is needed for a chat by applying the `@Timed` annotation to the `onMessage()` method.

See the {example-location}/src/main/java/dev/langchain4j/example/chat/ChatService.java[`src/main/java/dev/langchain4j/example/chat/ChatService.java`] file.
[source, java, role="no_copy"]
----
    @OnMessage
    @Timed(name = "chatProcessingTime",
           absolute = true,
           description = "Time needed chatting to the agent.")
    public void onMessage(String message, Session session) {
        ...
----

> And, collect how many agents are created by applying the `@Counted` annotation to the `AgentManager.createAgent()` method.

Reviewer: ```suggestion
It also collects how many agents are created by applying the `@Counted` annotation to the `AgentManager.createAgent()` method.
```
